core/domain/stats_jobs_one_off.py

```diff
@@ -16,16 +16,212 @@
+                stats_models.CompleteExplorationEventLogEntryModel
+               ]
+
+    @staticmethod
+    def map(item):
+        if item.event_schema_version == 2:
+            if isinstance(item, stats_models.StateCompleteEventLogEntryModel):
+                yield (item.exp_id,
+                       {
+                           'type': 'state_complete',
+                           'version': item.exp_version,
+                           'state': item.state_name
+                       })
+            elif isinstance(
+                    item, stats_models.AnswerSubmittedEventLogEntryModel):
+                yield (item.exp_id,
+                       {
+                           'type': 'answer_submitted',
+                           'version': item.exp_version,
+                           'state': item.state_name,
+                           'is_feedback_useful': item.is_feedback_useful
+                       })
+            elif isinstance(item, stats_models.StateHitEventLogEntryModel):
+                yield (item.exploration_id,
+                       {
+                           'type': 'state_hit',
+                           'version': item.exploration_version,
+                           'state': item.state_name,
+                           'session_id': item.session_id
+                       })
+            elif isinstance(item, stats_models.SolutionHitEventLogEntryModel):
+                yield (item.exp_id,
+                       {
+                           'type': 'solution_hit',
+                           'version': item.exp_version,
+                           'state': item.state_name,
+                           'session_id': item.session_id
+                       })
+            elif isinstance(
+                    item,
+                    stats_models.ExplorationActualStartEventLogEntryModel):
+                yield (item.exp_id,
+                       {
+                           'type': 'exploration_actual_start',
+                           'version': item.exp_version,
+                           'state': item.state_name
+                       })
+            elif isinstance(
+                    item, stats_models.CompleteExplorationEventLogEntryModel):
+                yield (item.exploration_id,
+                       {
+                           'type': 'complete_exploration',
+                           'version': item.exploration_version,
+                           'state': item.state_name
+                       })
+
+
+    @staticmethod
+    def reduce(key, values):
+        values = map(ast.literal_eval, values)
+        values = sorted(values, key=lambda x: x['version'])
+
+        for version, events in itertools.groupby(values,
+                                                 key=lambda x: x['version']):
+            if version == 1:
+                # Get a copy of the uncorrupted *_v1 statistics
+                old_stats = stats_services.get_exploration_stats_by_id(
```

Comment:
You're doing GET operations inside a for loop and that's way too many frequent datastore accessing. A better idea is to retrieve the ExplorationStatsModels for all versions at the outset - 
stats_services.get_multiple_exploration_stats_by_version method.

Reply:
Done.